discovery/reliable_sender.go

```diff
@@ -0,0 +1,316 @@
+	IsMsgStale func(lnwire.Message) bool
+}
+
+// peerManager contains the set of channels required for the peerHandler to
+// properly carry out its duties.
+type peerManager struct {
+	// msgs is the channel through which messages will be streamed to the
+	// handler in order to send the message to the peer while they're
+	// online.
+	msgs chan lnwire.Message
+
+	// done is a channel that will be closed to signal that the handler for
+	// the given peer has been torn down for whatever reason.
+	done chan struct{}
+}
+
+// reliableSender is a small subsystem of the gossiper used to reliably send
+// gossip messages to peers.
+type reliableSender struct {
+	start sync.Once
+	stop  sync.Once
+
+	cfg reliableSenderCfg
+
+	// activePeers keeps track of whether a peerHandler exists for a given
+	// peer. A peerHandler is tasked with handling requests for messages
+	// that should be reliably sent to peers while also taking into account
+	// the peer's connection lifecycle.
+	activePeers    map[[33]byte]peerManager
+	activePeersMtx sync.Mutex
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// newReliableSender returns a new reliableSender backed by the given config.
+func newReliableSender(cfg *reliableSenderCfg) *reliableSender {
+	return &reliableSender{
+		cfg:         *cfg,
+		activePeers: make(map[[33]byte]peerManager),
+		quit:        make(chan struct{}),
+	}
+}
+
+// Start spawns message handlers for any peers with pending messages.
+func (s *reliableSender) Start() error {
+	var err error
+	s.start.Do(func() {
+		err = s.resendPendingMsgs()
+	})
+	return err
+}
+
+// Stop halts the reliable sender from sending messages to peers.
+func (s *reliableSender) Stop() {
+	s.stop.Do(func() {
+		close(s.quit)
+		s.wg.Wait()
+	})
+}
+
+// sendMessage constructs a request to send a message reliably to a peer. In the
+// event that the peer is currently offline, this will only write the message to
+// disk. Once the peer reconnects, this message, along with any others pending,
+// will be sent to the peer.
+func (s *reliableSender) sendMessage(msg lnwire.Message, peerPubKey [33]byte) error {
+	// We'll start by persisting the message to disk. This allows us to
+	// resend the message upon restarts and peer reconnections.
+	if err := s.cfg.MessageStore.AddMessage(msg, peerPubKey); err != nil {
+		return err
+	}
+
+	// Then, we'll spawn a peerHandler for this peer to handle resending its
+	// pending messages while taking into account its connection lifecycle.
+spawnHandler:
+	msgHandler, ok := s.spawnPeerHandler(peerPubKey)
+
+	// If the handler wasn't previously active, we can exit now as we know
+	// that the message will be sent once the peer online notification is
+	// received. This prevents us from potentially sending the message
+	// twice.
+	if !ok {
+		return nil
+	}
+
+	// Otherwise, we'll attempt to stream the message to the handler.
+	// There's a subtle race condition where the handler can be torn down
```

Comment:
This is still a bit strange, but I guess it is okay 🙃 Current solution _should_ work.

Reply:
I don't think there are any other ways around it. Do you have any suggestions?